pandas/core/ops/array_ops.py

```diff
@@ -132,3 +152,177 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
         result = masked_arith_op(left, right, op)
 
     return missing.dispatch_fill_zeros(op, left, right, result)
+
+
+def arithmetic_op(left, right, op, str_rep, eval_kwargs):
+
+    from pandas.core.ops import (
+        maybe_upcast_for_op,
+        should_extension_dispatch,
+        dispatch_to_extension_op,
+    )
+
+    keep_null_freq = isinstance(
+        right,
+        (
+            ABCDatetimeIndex,
+            ABCDatetimeArray,
+            ABCTimedeltaIndex,
+            ABCTimedeltaArray,
+            Timestamp,
+        ),
+    )
+
+    # NB: We assume that extract_array has already been called on `left`, but
+    #  cannot make the same assumption about `right`.  This is because we need
+    #  to define `keep_null_freq` before calling extract_array on it.
+    lvalues = left
+    rvalues = extract_array(right, extract_numpy=True)
```

Comment:
but DatetimelikeArrays keep the `freq` ? So why is it not possible to extract up to the array level for those?

And for a scalar Timestamp, does this freq actually matter?

Reply:
> And for a scalar Timestamp, does this freq actually matter?

The behavior is deprecated, but `ts + 5` currently behaves like `ts + 5*ts.freq`, so yes, `freq` does matter. 

> but DatetimelikeArrays keep the freq ? So why is it not possible to extract up to the array level for those?

Suppose we have `self = Series(range(5))`, so `lvalues = np.array(range(5))`.  Consider two cases for `other` that we are going to add to `self`:

1) `other = DatetimeIndex(["2016-01-01"] * 5, tz="UTC")`
2) `other = Series(DatetimeIndex(["2016-01-01"] * 5, tz="UTC"))`

In the first case, `other` has a `freq` attribute that is None, so the correct thing to do is raise NullFrequencyError (btw this behavior is deprecated as with Timestamp).  In the second, there is no freq attribute, so the correct thing to do is raise TypeError.  If we did the unwrapping before getting to `arithmetic_op`, we would incorrectly raise NullFrequencyError in case 2.

Reply:
can you move *both* extract_array's in here? would make groking this simpler

Reply:
> But once the deprecation is removed, this will always be a type error? And then this does not need a special case here? (just for understanding, not expecting it be done in this PR ;))

Correct x 2.